stdlib/Random/src/RNGs.jl

```diff
@@ -172,6 +172,149 @@ function fillcache_zeros!(r::MersenneTwister)
+    s3 = s3 << 45 | s3 >> 19
+    rng.s0, rng.s1, rng.s2, rng.s3 = s0, s1, s2, s3 
+    res 
+end
+
+
+## Task local RNG
+"""
+    TaskLocal
+
+The TaskLocal RNG has state that is local to its task, not its thread. It is seeded upon task creation, from the state of its parent task. Therefore, task creation is an event that changes the parent's RNG state.
+
+As an upside, the TaskLocal RNG is pretty fast, and permits reproducible multithreaded simulations (barring race conditions), independent of scheduler decisions. As long as the number of threads is not used to make decisions on task creation, simulation results are also independent of the number of available threads / CPUs. The random stream should not depend on hardware specifics, up to endianness and possibly word size.
+
+Using or seeding the RNG of any other task than the one returned by `current_task()` is undefined behavior: it will work most of the time, and may sometimes fail silently.
+"""
+struct TaskLocal <: AbstractRNG end
+TaskLocal(::Nothing) = TaskLocal()
+rng_native_52(::TaskLocal) = UInt64
+
+function seed!(::TaskLocal, s0::UInt64, s1::UInt64, s2::UInt64, s3::UInt64)
+    task = current_task()
+    ccall(:jl_tasklocal_seedrandom, Nothing, (Ref{Task}, UInt64, UInt64, UInt64, UInt64), task, s0,s1,s2,s3)
+    TaskLocal()
+end
+
+
+@inline function rand(::TaskLocal, ::Type{UInt64})
+    task = current_task()
+    s0, s1, s2, s3 = task.rngState0, task.rngState1, task.rngState2, task.rngState3
+    tmp = s0 + s3
+    res = tmp << 23 | tmp >> 41
+    t = s1 << 17
+    s2 = xor(s2, s0)
+    s3 = xor(s3, s1)
+    s1 = xor(s1, s2)
+    s0 = xor(s0, s3)
+    s2 = xor(s2, t)
+    s3 = s3 << 45 | s3 >> 19
+    task.rngState0, task.rngState1, task.rngState2, task.rngState3 =  s0, s1, s2, s3
+    res 
+end
+
+#Shared implementation between Xoshiro and TaskLocal -- seeding
+function seed!(rng::Union{TaskLocal, Xoshiro}, seed::Union{Int128, UInt128, BigInt}) 
+    seed0 = seed % UInt64
+    seed1 = (seed>>64) % UInt64
+    seed!(rng, seed0, seed1, zero(UInt64), zero(UInt64))
+end
+seed!(rng::Union{TaskLocal, Xoshiro}, seed::Integer) = seed!(rng, seed%UInt64, zero(UInt64), zero(UInt64), zero(UInt64))
+seed!(rng::Union{TaskLocal, Xoshiro}, ::Nothing) = seed!(rng)
+
+seed!(rng::Union{TaskLocal, Xoshiro}) = seed!(rng,
+                           rand(RandomDevice(), UInt64),
+                           rand(RandomDevice(), UInt64),
+                           rand(RandomDevice(), UInt64),
+                           rand(RandomDevice(), UInt64))
+
+function seed!(rng::Union{TaskLocal, Xoshiro}, seed::AbstractVector{UInt64}) 
+    seed0 = length(seed)>0 ? seed[1] : UInt64(0)
+    seed1 = length(seed)>1 ? seed[2] : UInt64(0)
+    seed2 = length(seed)>2 ? seed[3] : UInt64(0)
+    seed3 = length(seed)>3 ? seed[4] : UInt64(0)
+    seed!(rng, seed0, seed1, seed2, seed3)
+end
+
+function seed!(rng::Union{TaskLocal, Xoshiro}, seed::AbstractVector{UInt32}) 
+    #can only process an even length
```

Comment:
Couldn't the last entry of an odd-lengthed vector be cast to an `UInt64` ? I guess that could be implemented later, but no bits of the input should be discarded (i.e. better to error out on odd length).